signer/xpi/cose.go

```diff
@@ -0,0 +1,259 @@
+		resultErr = fmt.Errorf("Expected SignMessage payload to be nil, but got %+v", msg.Payload)
+		return
+	}
+	if len(msg.Headers.Unprotected) != 0 {
+		resultErr = fmt.Errorf("Expected SignMessage Unprotected headers to be empty, but got %+v", msg.Headers.Unprotected)
+		return
+	}
+
+	if len(msg.Headers.Protected) != 1 {
+		resultErr = fmt.Errorf("Expected SignMessage Protected headers must contain one value, but got %d", len(msg.Headers.Protected))
+		return
+	}
+	kidValue, ok := msg.Headers.Protected[4] // 4 is the compressed key for "kid"
+	if !ok {
+		resultErr = fmt.Errorf("Expected SignMessage must have kid in Protected Headers")
+		return
+	}
+	// check that all kid values are bytes and decode into certs
+	kidArray, ok := kidValue.([]interface{})
+	if !ok {
+		resultErr = fmt.Errorf("Expected SignMessage Protected Headers kid value to be an array got %+v with type %T", kidValue, kidValue)
+		return
+	}
+	for i, cert := range kidArray {
+		certBytes, ok := cert.([]byte)
+		if !ok {
+			resultErr = fmt.Errorf("Expected SignMessage Protected Headers kid value %d to be a byte slice got %+v with type %T", i, cert, cert)
+			return
+		}
+		intermediateCert, err := x509.ParseCertificate(certBytes)
+		if err != nil {
+			resultErr = errors.Wrapf(err, "SignMessage Signature Protected Headers kid value %d does not decode to a parseable X509 cert", i)
+			return
+		}
+		intermediateCerts = append(intermediateCerts, intermediateCert)
+	}
+
+	for i, sig := range msg.Signatures {
+		eeCert, err := isValidCOSESignature(sig)
+		if err != nil {
+			resultErr = errors.Wrapf(err, "cose signature %d is invalid", i)
+			return
+		}
+		eeCerts = append(eeCerts, eeCert)
+	}
+
+	return
+}
+
+// verifyCOSESignatures checks that:
+//
+// 1) COSE manifest and signature files are present
+// 2) the PKCS7 manifest is present
+// 3) the COSE and PKCS7 manifests do not include COSE files
+// 4) we can decode the COSE signature and it has the right format for an XPI
+// 5) the right number of signatures are present and all intermediate and end entity certs parse properly
+// TODO: 6) there is a trusted path from the included COSE EE certs to the signer cert using the provided intermediates
+//
+func verifyCOSESignatures(signedFile signer.SignedFile, signOptions Options) error {
+	coseManifestBytes, err := readFileFromZIP(signedFile, "META-INF/cose.manifest")
+	if err != nil {
+		return fmt.Errorf("failed to read META-INF/cose.manifest from signed zip: %v", err)
+	}
+	coseMsgBytes, err := readFileFromZIP(signedFile, "META-INF/cose.sig")
+	if err != nil {
+		return fmt.Errorf("failed to read META-INF/cose.sig from signed zip: %v", err)
+	}
+	pkcs7ManifestBytes, err := readFileFromZIP(signedFile, "META-INF/manifest.mf")
+	if err != nil {
+		return fmt.Errorf("failed to read META-INF/manifest.mf from signed zip: %v", err)
+	}
+	coseManifest := string(coseManifestBytes)
+	pkcs7Manifest := string(pkcs7ManifestBytes)
+
+	if !strings.Contains(pkcs7Manifest, "cose") {
```

Comment:
couple things:
1. you could avoid switching to a string type by using `bytes.Contains(manifest, "\x63\x6f\x73\x65")`
2. manifests can be large and this is probably going to be slow, not sure if we care
3. this isn't exactly a deterministic match, any word containing `cose` will match, and [there are quite a few](https://www.thefreedictionary.com/words-containing-cose). I would suggest a deterministic match on the path in the manifest instead: `Name: META-INF/cose.sig` or `META-INF/cose.manifest`.

Reply:
This is only called from the go client, but it can't hurt to improve it.